<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Alexander Dell’Orti" />

<meta name="date" content="2025-01-27" />

<title>manaGENEment</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">manaGENEment</h1>
<h4 class="author">Alexander Dell’Orti</h4>
<h4 class="date">2025-01-27</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The manaGENEment package offers a framework for managing various gene
types using S4 class objects and corresponding methods. The package is
designed to handle protein-coding genes, long non-coding RNA (lncRNA)
genes, and microRNA (miRNA) genes, making the organization,
manipulation, and analysis of gene-related data within R easy. This
package utilizes GenomicRanges, IRanges, and Biostrings for genomic data
handling, making it easy to be integrated in the most common
pipelines.</p>
<div id="key-features" class="section level3">
<h3>Key Features</h3>
<ul>
<li>S4 Class Structures: Defines classes for different gene types with
specific attributes.</li>
<li>Accessor and Editor Methods: Provides functions to retrieve and
modify gene attributes.</li>
<li>Length Computation: Implements methods to calculate the length of
gene products based on gene type.</li>
</ul>
</div>
</div>
<div id="getting-started" class="section level2">
<h2>Getting Started</h2>
<p>Let’s see how to create gene objects and using their methods.</p>
</div>
<div id="creating-gene-objects" class="section level2">
<h2>Creating Gene Objects</h2>
<p>The core of the manaGENEment package revolves around its S4 classes:
Gene, ProteinCodingGene, lncRNAGene, and miRNAGene. For each of these
genes we have an RNA sequence, that in this version of the package, is
considered the final RNA sequence resulting from post-transcriptional
modifications. Consequently, entries with the same Gene ID and gene
sequence can have different RNA IDs and RNA sequences due to the
presence of different transcript isoforms. Below are examples of how to
instantiate objects for each of these classes:</p>
<div id="creating-a-protein-coding-gene-object" class="section level3">
<h3>Creating a Protein Coding Gene Object</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Create a ProteinCodingGene object</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>pcg <span class="ot">&lt;-</span> <span class="fu">ProteinCodingGene</span>(</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  <span class="at">geneID =</span> <span class="st">&quot;ENST000001&quot;</span>,</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>  <span class="at">h_symbol =</span> <span class="st">&quot;ACTB&quot;</span>,</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>  <span class="at">geneName =</span> <span class="st">&quot;Beta Actin&quot;</span>,</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>  <span class="at">description =</span> <span class="st">&quot;One of six different actin isoforms and a major component of the cytoskeleton&quot;</span>,</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>  <span class="at">geneStructure =</span> GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>    <span class="at">seqnames =</span> <span class="st">&quot;chr7&quot;</span>,</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>    <span class="at">ranges =</span> IRanges<span class="sc">::</span><span class="fu">IRanges</span>(<span class="at">start =</span> <span class="dv">5527175</span>, <span class="at">end =</span> <span class="dv">5535161</span>),</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>    <span class="at">strand =</span> <span class="st">&quot;+&quot;</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  ),</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>  <span class="at">RnaID =</span> <span class="st">&quot;ENST_RNA_001&quot;</span>,</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>  <span class="at">RnaSEQ =</span> Biostrings<span class="sc">::</span><span class="fu">RNAString</span>(<span class="st">&quot;AUGGCUAGCUAGCUAGCUA&quot;</span>),</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>  <span class="at">proteinID =</span> <span class="st">&quot;ENSP000001&quot;</span>,</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>  <span class="at">proteinSEQ =</span> Biostrings<span class="sc">::</span><span class="fu">AAString</span>(<span class="st">&quot;MEEEIAALVIDNGSG&quot;</span>)</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="creating-a-long-non-coding-rna-lncrna-gene-object" class="section level3">
<h3>Creating a Long Non-Coding RNA (lncRNA) Gene Object</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Create a lncRNAGene object</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>lnc <span class="ot">&lt;-</span> <span class="fu">lncRNAGene</span>(</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>  <span class="at">geneID =</span> <span class="st">&quot;ENST000002&quot;</span>,</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  <span class="at">h_symbol =</span> <span class="st">&quot;HOTAIR&quot;</span>,</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>  <span class="at">geneName =</span> <span class="st">&quot;HOX Transcript Antisense RNA&quot;</span>,</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>  <span class="at">description =</span> <span class="st">&quot;Involved in epigenetic regulation&quot;</span>,</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  <span class="at">geneStructure =</span> GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>    <span class="at">seqnames =</span> <span class="st">&quot;chr12&quot;</span>,</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>    <span class="at">ranges =</span> IRanges<span class="sc">::</span><span class="fu">IRanges</span>(<span class="at">start =</span> <span class="dv">1150000</span>, <span class="at">end =</span> <span class="dv">1160000</span>),</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>    <span class="at">strand =</span> <span class="st">&quot;-&quot;</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>  ),</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>  <span class="at">RnaID =</span> <span class="st">&quot;ENST_RNA_002&quot;</span>,</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>  <span class="at">RnaSEQ =</span> Biostrings<span class="sc">::</span><span class="fu">RNAString</span>(<span class="st">&quot;AUGGCUAGCUAGCUAGCUA&quot;</span>)</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="creating-a-microrna-mirna-gene-object" class="section level3">
<h3>Creating a microRNA (miRNA) Gene Object</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Create a miRNAGene object</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>mir <span class="ot">&lt;-</span> <span class="fu">miRNAGene</span>(</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">geneID =</span> <span class="st">&quot;ENST000003&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">h_symbol =</span> <span class="st">&quot;MIR21&quot;</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  <span class="at">geneName =</span> <span class="st">&quot;miR-21&quot;</span>,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>  <span class="at">description =</span> <span class="st">&quot;Oncogenic miRNA involved in various cancers&quot;</span>,</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>  <span class="at">geneStructure =</span> GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>    <span class="at">seqnames =</span> <span class="st">&quot;chr17&quot;</span>,</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>    <span class="at">ranges =</span> IRanges<span class="sc">::</span><span class="fu">IRanges</span>(<span class="at">start =</span> <span class="dv">43045629</span>, <span class="at">end =</span> <span class="dv">43045716</span>),</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>    <span class="at">strand =</span> <span class="st">&quot;+&quot;</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>  ),</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>  <span class="at">RnaID =</span> <span class="st">&quot;ENST_RNA_003&quot;</span>,</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>  <span class="at">RnaSEQ =</span> Biostrings<span class="sc">::</span><span class="fu">RNAString</span>(<span class="st">&quot;AUGGCUAGCUAGCUAGCUA&quot;</span>),</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>  <span class="at">mirna_seed_SEQ =</span> Biostrings<span class="sc">::</span><span class="fu">RNAString</span>(<span class="st">&quot;UAGCUA&quot;</span>)</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="gene-objects-methods" class="section level2">
<h2>Gene Objects Methods</h2>
<p>The manaGENEment package provides a set of generic functions and
methods to interact with gene objects. These include accessor functions
to retrieve gene attributes and editing functions to modify them.</p>
<div id="accessor-functions" class="section level3">
<h3>Accessor Functions</h3>
<p><strong>Getting Gene ID</strong></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Retrieve the Gene ID from the ProteinCodingGene object</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>gene_id_pcg <span class="ot">&lt;-</span> <span class="fu">getGeneID</span>(pcg)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">print</span>(gene_id_pcg)</span></code></pre></div>
<pre><code>#&gt; [1] &quot;ENST000001&quot;</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Retrieve the Gene ID from the lncRNAGene object</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>gene_id_lnc <span class="ot">&lt;-</span> <span class="fu">getGeneID</span>(lnc)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">print</span>(gene_id_lnc)</span></code></pre></div>
<pre><code>#&gt; [1] &quot;ENST000002&quot;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Retrieve the Gene ID from the miRNAGene object</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>gene_id_mir <span class="ot">&lt;-</span> <span class="fu">getGeneID</span>(mir)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">print</span>(gene_id_mir)</span></code></pre></div>
<pre><code>#&gt; [1] &quot;ENST000003&quot;</code></pre>
<p><strong>Getting RNA Sequence</strong></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Retrieve the RNA ID from the ProteinCodingGene object</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>rna_id_pcg <span class="ot">&lt;-</span> <span class="fu">getRnaID</span>(pcg)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="fu">print</span>(rna_id_pcg)</span></code></pre></div>
<pre><code>#&gt; [1] &quot;ENST_RNA_001&quot;</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># Retrieve the RNA ID from the lncRNAGene object</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>rna_id_lnc <span class="ot">&lt;-</span> <span class="fu">getRnaID</span>(lnc)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="fu">print</span>(rna_id_lnc)</span></code></pre></div>
<pre><code>#&gt; [1] &quot;ENST_RNA_002&quot;</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Retrieve the RNA ID from the miRNAGene object</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>rna_id_mir <span class="ot">&lt;-</span> <span class="fu">getRnaID</span>(mir)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="fu">print</span>(rna_id_mir)</span></code></pre></div>
<pre><code>#&gt; [1] &quot;ENST_RNA_003&quot;</code></pre>
</div>
<div id="manipulation-functions" class="section level3">
<h3>Manipulation Functions</h3>
<p><strong>Setting Gene ID</strong></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># Set a new Gene ID for the ProteinCodingGene object</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>pcg <span class="ot">&lt;-</span> <span class="fu">setGeneID</span>(pcg, <span class="st">&quot;ENST000004&quot;</span>)</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">getGeneID</span>(pcg))</span></code></pre></div>
<pre><code>#&gt; [1] &quot;ENST000004&quot;</code></pre>
</div>
<div id="retrieve-length-of-the-product" class="section level3">
<h3>Retrieve length of the product</h3>
<p>The lengthProduct method computes the length of the gene’s main
product, which varies based on the gene type.</p>
<p><strong>For Protein Coding Genes</strong></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="co"># Compute the length of the protein sequence</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>len_pcg <span class="ot">&lt;-</span> <span class="fu">lengthProduct</span>(pcg)</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="fu">print</span>(len_pcg)  <span class="co"># Number of amino acids</span></span></code></pre></div>
<pre><code>#&gt; [1] 15</code></pre>
<p><strong>For Long Non-Coding RNA Genes</strong></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="co"># Compute the length of the RNA sequence</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>len_lnc <span class="ot">&lt;-</span> <span class="fu">lengthProduct</span>(lnc)</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="fu">print</span>(len_lnc)  <span class="co"># Length of RNA sequence</span></span></code></pre></div>
<pre><code>#&gt; [1] 19</code></pre>
<p><strong>For microRNA Genes</strong></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># Compute the length of the RNA sequence</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>len_mir <span class="ot">&lt;-</span> <span class="fu">lengthProduct</span>(mir)</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="fu">print</span>(len_mir)  <span class="co"># Length of RNA sequence</span></span></code></pre></div>
<pre><code>#&gt; [1] 19</code></pre>
</div>
</div>
<div id="specialized-tools" class="section level2">
<h2>Specialized tools</h2>
<p>Other than basic creation and manipulation, you can access
information that is specific and exclusive of a gene type</p>
<p><strong>Retrieving Protein Information from Protein Coding
Genes</strong></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="co"># Retrieve the Protein ID and Sequence from the ProteinCodingGene object</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>prot_id <span class="ot">&lt;-</span> <span class="fu">getProtID</span>(pcg)</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>prot_seq <span class="ot">&lt;-</span> <span class="fu">getProtSeq</span>(pcg)</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a><span class="fu">print</span>(prot_id)</span></code></pre></div>
<pre><code>#&gt; [1] &quot;ENSP000001&quot;</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="fu">print</span>(prot_seq)</span></code></pre></div>
<pre><code>#&gt; 15-letter AAString object
#&gt; seq: MEEEIAALVIDNGSG</code></pre>
<p><strong>Retrieving micro RNA Seed Sequence</strong></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="co"># Retrieve the miRNA seed sequence from the miRNAGene object</span></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>seed_seq <span class="ot">&lt;-</span> <span class="fu">getSeedSEQ</span>(mir)</span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a><span class="fu">print</span>(seed_seq)</span></code></pre></div>
<pre><code>#&gt; 6-letter RNAString object
#&gt; seq: UAGCUA</code></pre>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<ul>
<li><strong>GenomicRanges</strong> : A package for representing and
manipulating genomic intervals and locations.</li>
<li><strong>IRanges</strong> : Package for representing and manipulating
ranges.</li>
<li><strong>Biostrings</strong> : Package for representation and
manipulation of biological strings such as DNA, RNA, and AA
sequences.</li>
</ul>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
