---
title: "manaGENEment"
author: "Alexander Dell'Orti"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{manaGENEment}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, echo=FALSE}
# Load the manaGENEment package
library(manaGENEment)
```


## Introduction

The manaGENEment package offers a framework for managing various gene types using S4 class objects and corresponding methods. The package is designed to handle protein-coding genes, long non-coding RNA (lncRNA) genes, and microRNA (miRNA) genes, making the organization, manipulation, and analysis of gene-related data within R easy.
This package utilizes GenomicRanges, IRanges, and Biostrings for genomic data handling, making it easy to be integrated in the most common pipelines.

### Key Features
* S4 Class Structures: Defines classes for different gene types with specific attributes.
* Accessor and Editor Methods: Provides functions to retrieve and modify gene attributes.
* Length Computation: Implements methods to calculate the length of gene products based on gene type.


## Getting Started
Let's see how to create gene objects and using their methods.

## Creating Gene Objects

The core of the manaGENEment package revolves around its S4 classes: Gene, ProteinCodingGene, lncRNAGene, and miRNAGene.
For each of these genes we have an RNA sequence, that in this version of the package, is considered the final RNA sequence resulting from post-transcriptional modifications. Consequently, entries with the same Gene ID and gene sequence can have different RNA IDs and RNA sequences due to the presence of different transcript isoforms.
Below are examples of how to instantiate objects for each of these classes:

### Creating a Protein Coding Gene Object
```{r, echo=TRUE, collapse =FALSE}

# Create a ProteinCodingGene object
pcg <- ProteinCodingGene(
  geneID = "ENST000001",
  h_symbol = "ACTB",
  geneName = "Beta Actin",
  description = "One of six different actin isoforms and a major component of the cytoskeleton",
  geneStructure = GenomicRanges::GRanges(
    seqnames = "chr7",
    ranges = IRanges::IRanges(start = 5527175, end = 5535161),
    strand = "+"
  ),
  RnaID = "ENST_RNA_001",
  RnaSEQ = Biostrings::RNAString("AUGGCUAGCUAGCUAGCUA"),
  proteinID = "ENSP000001",
  proteinSEQ = Biostrings::AAString("MEEEIAALVIDNGSG")
)

```


### Creating a Long Non-Coding RNA (lncRNA) Gene Object
```{r, echo=TRUE, collapse =FALSE}
# Create a lncRNAGene object
lnc <- lncRNAGene(
  geneID = "ENST000002",
  h_symbol = "HOTAIR",
  geneName = "HOX Transcript Antisense RNA",
  description = "Involved in epigenetic regulation",
  geneStructure = GenomicRanges::GRanges(
    seqnames = "chr12",
    ranges = IRanges::IRanges(start = 1150000, end = 1160000),
    strand = "-"
  ),
  RnaID = "ENST_RNA_002",
  RnaSEQ = Biostrings::RNAString("AUGGCUAGCUAGCUAGCUA")
)

```



### Creating a microRNA (miRNA) Gene Object
```{r, echo=TRUE, collapse =FALSE}
# Create a miRNAGene object
mir <- miRNAGene(
  geneID = "ENST000003",
  h_symbol = "MIR21",
  geneName = "miR-21",
  description = "Oncogenic miRNA involved in various cancers",
  geneStructure = GenomicRanges::GRanges(
    seqnames = "chr17",
    ranges = IRanges::IRanges(start = 43045629, end = 43045716),
    strand = "+"
  ),
  RnaID = "ENST_RNA_003",
  RnaSEQ = Biostrings::RNAString("AUGGCUAGCUAGCUAGCUA"),
  mirna_seed_SEQ = Biostrings::RNAString("UAGCUA")
)
```



## Gene Objects Methods
The manaGENEment package provides a set of generic functions and methods to interact with gene objects. These include accessor functions to retrieve gene attributes and editing functions to modify them.

### Accessor Functions
__Getting Gene ID__
```{r, echo=TRUE, collapse =FALSE}
# Retrieve the Gene ID from the ProteinCodingGene object
gene_id_pcg <- getGeneID(pcg)
print(gene_id_pcg)

# Retrieve the Gene ID from the lncRNAGene object
gene_id_lnc <- getGeneID(lnc)
print(gene_id_lnc)

# Retrieve the Gene ID from the miRNAGene object
gene_id_mir <- getGeneID(mir)
print(gene_id_mir)

```


__Getting RNA Sequence__
```{r, echo=TRUE, collapse =FALSE}
# Retrieve the RNA ID from the ProteinCodingGene object
rna_id_pcg <- getRnaID(pcg)
print(rna_id_pcg)

# Retrieve the RNA ID from the lncRNAGene object
rna_id_lnc <- getRnaID(lnc)
print(rna_id_lnc)

# Retrieve the RNA ID from the miRNAGene object
rna_id_mir <- getRnaID(mir)
print(rna_id_mir)

```



### Manipulation Functions

__Setting Gene ID__
```{r, echo=TRUE, collapse =FALSE}
# Set a new Gene ID for the ProteinCodingGene object
pcg <- setGeneID(pcg, "ENST000004")
print(getGeneID(pcg))

```


### Retrieve length of the product
The lengthProduct method computes the length of the geneâ€™s main product, which varies based on the gene type.

__For Protein Coding Genes__
```{r, echo=TRUE, collapse =FALSE}
# Compute the length of the protein sequence
len_pcg <- lengthProduct(pcg)
print(len_pcg)  # Number of amino acids
```



__For Long Non-Coding RNA Genes__
```{r, echo=TRUE, collapse =FALSE}
# Compute the length of the RNA sequence
len_lnc <- lengthProduct(lnc)
print(len_lnc)  # Length of RNA sequence
```




__For microRNA Genes__
```{r, echo=TRUE, collapse =FALSE}
# Compute the length of the RNA sequence
len_mir <- lengthProduct(mir)
print(len_mir)  # Length of RNA sequence
```



## Specialized tools
Other than basic creation and manipulation, you can access information that is specific and exclusive of a gene type 


__Retrieving Protein Information from Protein Coding Genes__
```{r, echo=TRUE, collapse =FALSE}
# Retrieve the Protein ID and Sequence from the ProteinCodingGene object
prot_id <- getProtID(pcg)
prot_seq <- getProtSeq(pcg)
print(prot_id)
print(prot_seq)
```




__Retrieving micro RNA Seed Sequence__
```{r, echo=TRUE, collapse =FALSE}
# Retrieve the miRNA seed sequence from the miRNAGene object
seed_seq <- getSeedSEQ(mir)
print(seed_seq)
```






## References
* __GenomicRanges__ : A package for representing and manipulating genomic intervals and locations.
* __IRanges__ : Package for representing and manipulating ranges.
* __Biostrings__ : Package for representation and manipulation of biological strings such as DNA, RNA, and AA sequences.
